File: packages/core/src/derive/verdict.ts
import type { EvidenceSource, EvidenceWaiver, Polarity } from "../schema/evidence.ts";
import type { SourceCondition, VerdictPolicy } from "../schema/policy.ts";

/**
 * What one evidence channel actually showed.
 *
 * Kept to a small set of facts rather than raw captures so that the verdict
 * rules stay checkable: whoever collected the evidence decides what counts as
 * an error on their channel, and this module decides what that means for the
 * outcome.
 */
export interface Observation {
  readonly source: EvidenceSource;
  /**
   * Whether the channel was successfully read.
   *
   * False is not the same as "nothing was wrong". Evidence that was supposed
   * to be gathered and was not leaves the verdict unsettled, and saying so is
   * the whole reason this field exists.
   */
  readonly collected: boolean;
  /** Errors, exceptions or violations seen on this channel. */
  readonly errors?: readonly string[];
  /** HTTP status observed, where the channel carries one. */
  readonly status?: number;
  /** Whether what was seen matches what the case said to expect. */
  readonly matchedExpectation?: boolean;
  /** Rows changed as the case intended. */
  readonly intendedChanges?: number;
  /** Rows changed that the case did not intend, including failed rollback. */
  readonly residualChanges?: number;
}

/** The outcome of judging a run against a policy. */
export type Verdict =
  /** Every binding condition was met. */
  | "pass"
  /** At least one binding condition was contradicted. */
  | "fail"
  /** Not enough evidence to decide. Never reported as a pass. */
  | "inconclusive";

/** Why a single source did not meet its condition. */
export interface ConditionFailure {
  readonly source: EvidenceSource;
  readonly condition: SourceCondition;
  readonly why: string;
}

/** The result of judging a run, with its reasoning intact. */
export interface VerdictResult {
  readonly verdict: Verdict;
  /** Conditions that were contradicted. */
  readonly failures: readonly ConditionFailure[];
  /** Binding sources that were never collected. */
  readonly missing: readonly EvidenceSource[];
  /** Sources that actually carried the decision. */
  readonly decidedBy: readonly EvidenceSource[];
  /**
   * Sources this case was excused from, and why.
   *
   * Reported rather than silently dropped: a verdict reached on fewer
   * channels is a weaker verdict, and the reviewer view says so.
   */
  readonly waived: readonly EvidenceWaiver[];
}

function checkCondition(condition: SourceCondition, observation: Observation): string | undefined {
  const errors = observation.errors ?? [];
  switch (condition) {
    case "informational":
      return undefined;
    case "clean":
      return errors.length === 0 ? undefined : `${errors.length} error(s) reported`;
    case "expected_error":
      if (errors.length === 0) return "no error was reported, but one was expected";
      return observation.matchedExpectation === true
        ? undefined
        : "an error was reported but it is not the expected one";
    case "matches_expectation":
      return observation.matchedExpectation === true
        ? undefined
        : "does not match the stated expectation";
    case "success_response": {
      const status = observation.status;
      if (status === undefined) return "no response status was observed";
      return status >= 200 && status < 300 ? undefined : `status ${status} is not a success`;
    }
    case "error_response": {
      const status = observation.status;
      if (status === undefined) return "no response status was observed";
      return status >= 400 ? undefined : `status ${status} is not a rejection`;
    }
    case "expected_change":
      return (observation.intendedChanges ?? 0) > 0
        ? undefined
        : "the intended data change was not observed";
    case "no_residual_change":
      return (observation.residualChanges ?? 0) === 0
        ? undefined
        : `${observation.residualChanges} unintended change(s) remain`;
  }
}

/**
 * Judges collected evidence against a policy.
 *
 * Three properties are deliberate:
 *
 * - **Missing evidence never passes.** A binding source that was not
 *   collected yields `inconclusive`, so a broken collector degrades into a
 *   visible gap rather than a green run.
 * - **Weak evidence never passes alone.** If the only sources that decided
 *   the outcome are ones the policy marks insufficient by themselves — a
 *   screenshot, typically — the result is `inconclusive`. This is what stops
 *   a suite from certifying a system whose backend failed quietly behind a
 *   correct-looking page.
 * - **Failures name their channel.** The caller learns which observation
 *   contradicted which condition, which is what makes triage possible.
 *
 * @param policy - Rules to judge by.
 * @param polarity - Whether the case expected success or rejection.
 * @param observations - What each channel showed.
 * @returns The verdict together with the reasoning behind it.
 */
export function evaluateVerdict(
  policy: VerdictPolicy,
  polarity: Polarity,
  observations: readonly Observation[],
  waivers: readonly EvidenceWaiver[] = [],
): VerdictResult {
  const rules = policy.rules[polarity] ?? {};
  const seen = new Map(observations.map((observation) => [observation.source, observation]));
  const waivedBySource = new Map(waivers.map((waiver) => [waiver.source, waiver]));

  const failures: ConditionFailure[] = [];
  const missing: EvidenceSource[] = [];
  const decidedBy: EvidenceSource[] = [];
  const waived: EvidenceWaiver[] = [];

  for (const [rawSource, rawCondition] of Object.entries(rules)) {
    const source = rawSource as EvidenceSource;
    const condition = rawCondition as SourceCondition;
    if (condition === "informational") continue;

    const waiver = waivedBySource.get(source);
    if (waiver) {
      if (policy.nonWaivable.includes(source)) {
        failures.push({
          source,
          condition,
          why: `this channel cannot be waived: ${waiver.reason}`,
        });
      } else {
        waived.push(waiver);
      }
      continue;
    }

    const observation = seen.get(source);
    if (!observation || !observation.collected) {
      missing.push(source);
      continue;
    }

    decidedBy.push(source);
    const why = checkCondition(condition, observation);
    if (why !== undefined) failures.push({ source, condition, why });
  }

  if (failures.length > 0) {
    return { verdict: "fail", failures, missing, decidedBy, waived };
  }
  if (missing.length > 0) {
    return { verdict: "inconclusive", failures, missing, decidedBy, waived };
  }

  const weakOnly =
    decidedBy.length > 0 && decidedBy.every((source) => policy.insufficientAlone.includes(source));
  if (decidedBy.length === 0 || weakOnly) {
    return { verdict: "inconclusive", failures, missing, decidedBy, waived };
  }

  return { verdict: "pass", failures, missing, decidedBy, waived };
}

/** A waiver a policy does not permit. */
export interface WaiverProblem {
  readonly source: EvidenceSource;
  readonly reason: string;
}

/**
 * Checks a case's waivers against a policy, for validation ahead of any run.
 *
 * Catching an impermissible waiver when the specification is saved is worth
 * more than catching it when the suite runs: the person who wrote it is still
 * there, and no run has yet been reported on evidence the team had already
 * decided was mandatory.
 *
 * @param policy - Policy the case will be judged under.
 * @param waivers - Waivers the case declares.
 * @returns The waivers the policy forbids; empty when all are permitted.
 */
export function validateWaivers(
  policy: VerdictPolicy,
  waivers: readonly EvidenceWaiver[],
): WaiverProblem[] {
  return waivers
    .filter((waiver) => policy.nonWaivable.includes(waiver.source))
    .map((waiver) => ({ source: waiver.source, reason: waiver.reason }));
}

File: python/agentic-test-hub-runner/src/agentic_test_hub_runner/verdict.py
"""
Judging collected evidence against a policy, mirroring
`packages/core/src/derive/verdict.ts`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from .policy import VerdictPolicy

EvidenceSource = Literal[
    "screenshot", "browser_console", "browser_network", "db_records", "app_log", "db_log"
]
Polarity = Literal["nominal", "error"]
SourceCondition = Literal[
    "clean",
    "expected_error",
    "matches_expectation",
    "success_response",
    "error_response",
    "expected_change",
    "no_residual_change",
    "informational",
]
Verdict = Literal["pass", "fail", "inconclusive"]

_DEFAULT_INSUFFICIENT_ALONE = ["screenshot"]
_DEFAULT_NON_WAIVABLE = ["browser_console", "app_log", "db_records"]


@dataclass(frozen=True)
class Observation:
    """What one evidence channel actually showed."""

    source: EvidenceSource
    collected: bool
    errors: list[str] = field(default_factory=list)
    status: int | None = None
    matched_expectation: bool | None = None
    intended_changes: int | None = None
    residual_changes: int | None = None
    suppressed: int | None = None


@dataclass(frozen=True)
class EvidenceWaiver:
    source: EvidenceSource
    reason: str


@dataclass(frozen=True)
class ConditionFailure:
    """Why a single source did not meet its condition."""

    source: EvidenceSource
    condition: SourceCondition
    why: str


@dataclass(frozen=True)
class VerdictResult:
    """The result of judging a run, with its reasoning intact."""

    verdict: Verdict
    failures: list[ConditionFailure] = field(default_factory=list)
    missing: list[EvidenceSource] = field(default_factory=list)
    decided_by: list[EvidenceSource] = field(default_factory=list)
    waived: list[EvidenceWaiver] = field(default_factory=list)


def _check_condition(condition: SourceCondition, observation: Observation) -> str | None:
    errors = observation.errors
    if condition == "informational":
        return None
    if condition == "clean":
        return None if len(errors) == 0 else f"{len(errors)} error(s) reported"
    if condition == "expected_error":
        if len(errors) == 0:
            return "no error was reported, but one was expected"
        return (
            None
            if observation.matched_expectation is True
            else "an error was reported but it is not the expected one"
        )
    if condition == "matches_expectation":
        return (
            None
            if observation.matched_expectation is True
            else "does not match the stated expectation"
        )
    if condition == "success_response":
        status = observation.status
        if status is None:
            return "no response status was observed"
        return None if 200 <= status < 300 else f"status {status} is not a success"
    if condition == "error_response":
        status = observation.status
        if status is None:
            return "no response status was observed"
        return None if status >= 400 else f"status {status} is not a rejection"
    if condition == "expected_change":
        return (
            None
            if (observation.intended_changes or 0) > 0
            else "the intended data change was not observed"
        )
    if condition == "no_residual_change":
        residual = observation.residual_changes or 0
        return None if residual == 0 else f"{residual} unintended change(s) remain"
    raise ValueError(f"unknown source condition: {condition}")


def evaluate_verdict(
    policy: VerdictPolicy,
    polarity: Polarity,
    observations: list[Observation],
    waivers: list[EvidenceWaiver] | None = None,
) -> VerdictResult:
    """
    Judges collected evidence against a policy.

    Three properties, same as the TypeScript original: missing evidence never
    passes (a binding source not collected yields `inconclusive`); weak
    evidence never passes alone (if every deciding source is one the policy
    marks `insufficientAlone`, the result is `inconclusive`); failures name
    their channel.
    """
    waivers = waivers or []
    rules: dict[EvidenceSource, SourceCondition] = policy.get("rules", {}).get(polarity, {})
    seen = {observation.source: observation for observation in observations}
    waived_by_source = {waiver.source: waiver for waiver in waivers}
    non_waivable = policy.get("nonWaivable", _DEFAULT_NON_WAIVABLE)
    insufficient_alone = policy.get("insufficientAlone", _DEFAULT_INSUFFICIENT_ALONE)

    failures: list[ConditionFailure] = []
    missing: list[EvidenceSource] = []
    decided_by: list[EvidenceSource] = []
    waived: list[EvidenceWaiver] = []

    for source, condition in rules.items():
        if condition == "informational":
            continue

        waiver = waived_by_source.get(source)
        if waiver is not None:
            if source in non_waivable:
                failures.append(
                    ConditionFailure(
                        source, condition, f"this channel cannot be waived: {waiver.reason}"
                    )
                )
            else:
                waived.append(waiver)
            continue

        observation = seen.get(source)
        if observation is None or not observation.collected:
            missing.append(source)
            continue

        decided_by.append(source)
        why = _check_condition(condition, observation)
        if why is not None:
            failures.append(ConditionFailure(source, condition, why))

    if failures:
        return VerdictResult("fail", failures, missing, decided_by, waived)
    if missing:
        return VerdictResult("inconclusive", failures, missing, decided_by, waived)

    weak_only = len(decided_by) > 0 and all(source in insufficient_alone for source in decided_by)
    if len(decided_by) == 0 or weak_only:
        return VerdictResult("inconclusive", failures, missing, decided_by, waived)

    return VerdictResult("pass", failures, missing, decided_by, waived)
